File: skills/agent-patterns/templates/go/evaluator_optimizer.go
/*
 * Evaluator-Optimizer Pattern Implementation for Go
 * Iterative refinement with generator and evaluator loop
 */

package agentpatterns

import (
	"context"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// EvaluationCriterion represents an evaluation criterion with weight
type EvaluationCriterion struct {
	Name        string
	Description string
	Weight      float64
}

// EvaluationResult represents the result of an evaluation
type EvaluationResult struct {
	OverallScore   float64
	CriteriaScores map[string]float64
	Feedback       string
	Suggestions    []string
}

// IterationRecord represents a record of an iteration
type IterationRecord struct {
	Iteration  int
	Output     string
	Evaluation *EvaluationResult
}

// EvaluatorOptimizer iteratively refines output.
//
// Example:
//
//	optimizer := NewEvaluatorOptimizer(client, "claude-sonnet-4-20250514")
//	optimizer.AddCriterion(EvaluationCriterion{Name: "clarity", Description: "Clear writing", Weight: 1.5})
//	result, err := optimizer.Optimize(ctx, "Write a blog post about AI", 3, 0.85)
type EvaluatorOptimizer struct {
	client         *AnthropicClient
	generatorModel string
	evaluatorModel string
	criteria       []EvaluationCriterion
	history        []IterationRecord
}

// NewEvaluatorOptimizer creates a new EvaluatorOptimizer
func NewEvaluatorOptimizer(client *AnthropicClient, model string) *EvaluatorOptimizer {
	return &EvaluatorOptimizer{
		client:         client,
		generatorModel: model,
		evaluatorModel: model,
		criteria:       []EvaluationCriterion{},
		history:        []IterationRecord{},
	}
}

// WithEvaluatorModel sets a different model for evaluation
func (e *EvaluatorOptimizer) WithEvaluatorModel(model string) *EvaluatorOptimizer {
	e.evaluatorModel = model
	return e
}

// AddCriterion adds an evaluation criterion
func (e *EvaluatorOptimizer) AddCriterion(criterion EvaluationCriterion) *EvaluatorOptimizer {
	e.criteria = append(e.criteria, criterion)
	return e
}

// History returns the iteration history
func (e *EvaluatorOptimizer) History() []IterationRecord {
	return e.history
}

// OptimizationResult represents the result of optimization
type OptimizationResult struct {
	FinalOutput  string
	FinalScore   float64
	Iterations   int
	MetThreshold bool
	History      []IterationRecord
}

// Optimize optimizes output through iterative refinement
func (e *EvaluatorOptimizer) Optimize(ctx context.Context, task string, maxIterations int, scoreThreshold float64) (*OptimizationResult, error) {
	e.history = []IterationRecord{}
	currentOutput := ""
	var lastEvaluation *EvaluationResult

	for i := 0; i < maxIterations; i++ {
		// Generate (or refine) output
		output, err := e.generate(ctx, task, currentOutput, lastEvaluation)
		if err != nil {
			return nil, fmt.Errorf("generation failed: %w", err)
		}
		currentOutput = output

		// Evaluate output
		evaluation, err := e.evaluate(ctx, currentOutput)
		if err != nil {
			return nil, fmt.Errorf("evaluation failed: %w", err)
		}

		// Record iteration
		e.history = append(e.history, IterationRecord{
			Iteration:  i + 1,
			Output:     currentOutput,
			Evaluation: evaluation,
		})

		// Check if we've met the threshold
		if evaluation.OverallScore >= scoreThreshold {
			return &OptimizationResult{
				FinalOutput:  currentOutput,
				FinalScore:   evaluation.OverallScore,
				Iterations:   i + 1,
				MetThreshold: true,
				History:      e.history,
			}, nil
		}

		lastEvaluation = evaluation
	}

	// Return best result after max iterations
	var bestIteration *IterationRecord
	var bestScore float64
	for i := range e.history {
		if e.history[i].Evaluation.OverallScore > bestScore {
			bestScore = e.history[i].Evaluation.OverallScore
			bestIteration = &e.history[i]
		}
	}

	return &OptimizationResult{
		FinalOutput:  bestIteration.Output,
		FinalScore:   bestIteration.Evaluation.OverallScore,
		Iterations:   maxIterations,
		MetThreshold: false,
		History:      e.history,
	}, nil
}

func (e *EvaluatorOptimizer) generate(ctx context.Context, task, previousOutput string, previousEvaluation *EvaluationResult) (string, error) {
	var prompt string

	if previousOutput == "" {
		prompt = fmt.Sprintf(`Complete this task:

%s

Provide your best output:`, task)
	} else {
		var feedbackText string
		if previousEvaluation != nil {
			var suggestions []string
			for _, s := range previousEvaluation.Suggestions {
				suggestions = append(suggestions, "- "+s)
			}
			feedbackText = fmt.Sprintf(`Previous evaluation feedback:
%s

Specific suggestions:
%s`, previousEvaluation.Feedback, strings.Join(suggestions, "\n"))
		}

		prompt = fmt.Sprintf(`Improve this output based on the feedback:

Original task: %s

Previous output:
%s

%s

Provide an improved version:`, task, previousOutput, feedbackText)
	}

	return e.client.CreateMessage(ctx, prompt, e.generatorModel, 4096)
}

func (e *EvaluatorOptimizer) evaluate(ctx context.Context, output string) (*EvaluationResult, error) {
	var criteriaList string
	if len(e.criteria) > 0 {
		var parts []string
		for _, c := range e.criteria {
			parts = append(parts, fmt.Sprintf("- %s (weight: %.1f): %s", c.Name, c.Weight, c.Description))
		}
		criteriaList = strings.Join(parts, "\n")
	} else {
		criteriaList = `- quality: Overall quality and correctness
- clarity: Clear and understandable
- completeness: Addresses all aspects`
	}

	prompt := fmt.Sprintf(`Evaluate this output against the following criteria:

%s

Output to evaluate:
%s

Respond with JSON in this exact format:
{
    "overall_score": 0.0-1.0,
    "criteria_scores": {
        "criterion_name": 0.0-1.0
    },
    "feedback": "Overall assessment",
    "suggestions": ["specific improvement 1", "specific improvement 2"]
}`, criteriaList, output)

	response, err := e.client.CreateMessage(ctx, prompt, e.evaluatorModel, 1024)
	if err != nil {
		return nil, err
	}

	return parseEvaluationJSON(response)
}

func parseEvaluationJSON(jsonStr string) (*EvaluationResult, error) {
	result := &EvaluationResult{
		OverallScore:   0.5,
		CriteriaScores: make(map[string]float64),
		Suggestions:    []string{},
	}

	// Extract overall score
	scoreRe := regexp.MustCompile(`"overall_score"\s*:\s*([0-9.]+)`)
	if match := scoreRe.FindStringSubmatch(jsonStr); len(match) > 1 {
		if score, err := strconv.ParseFloat(match[1], 64); err == nil {
			result.OverallScore = score
		}
	}

	// Extract feedback
	feedbackRe := regexp.MustCompile(`"feedback"\s*:\s*"([^"]*)"`)
	if match := feedbackRe.FindStringSubmatch(jsonStr); len(match) > 1 {
		result.Feedback = match[1]
	}

	// Extract suggestions
	suggestionsRe := regexp.MustCompile(`"suggestions"\s*:\s*\[(.*?)\]`)
	if match := suggestionsRe.FindStringSubmatch(jsonStr); len(match) > 1 {
		suggestionItemRe := regexp.MustCompile(`"([^"]+)"`)
		items := suggestionItemRe.FindAllStringSubmatch(match[1], -1)
		for _, item := range items {
			if len(item) > 1 {
				result.Suggestions = append(result.Suggestions, item[1])
			}
		}
	}

	return result, nil
}

// ConfidenceBasedOptimizer generates with confidence self-assessment
type ConfidenceBasedOptimizer struct {
	client *AnthropicClient
	model  string
}

// NewConfidenceBasedOptimizer creates a new ConfidenceBasedOptimizer
func NewConfidenceBasedOptimizer(client *AnthropicClient, model string) *ConfidenceBasedOptimizer {
	return &ConfidenceBasedOptimizer{
		client: client,
		model:  model,
	}
}

// AttemptRecord represents a record of an attempt
type AttemptRecord struct {
	Attempt    int
	Output     string
	Confidence float64
}

// ConfidenceResult represents the result of confidence-based generation
type ConfidenceResult struct {
	Output       string
	Confidence   float64
	Attempts     []AttemptRecord
	MetThreshold bool
}

// GenerateWithConfidence generates with confidence self-assessment
func (c *ConfidenceBasedOptimizer) GenerateWithConfidence(ctx context.Context, task string, confidenceThreshold float64, maxAttempts int) (*ConfidenceResult, error) {
	var attempts []AttemptRecord
	bestOutput := ""
	bestConfidence := 0.0

	for i := 0; i < maxAttempts; i++ {
		prompt := fmt.Sprintf(`Complete this task and assess your confidence:

%s

After your response, on a new line, provide your confidence level (0.0-1.0) that your answer is correct and complete.

Format:
[Your response here]

CONFIDENCE: [0.0-1.0]`, task)

		response, err := c.client.CreateMessage(ctx, prompt, c.model, 4096)
		if err != nil {
			return nil, err
		}

		output, confidence := parseConfidenceResponse(response)

		attempts = append(attempts, AttemptRecord{
			Attempt:    i + 1,
			Output:     output,
			Confidence: confidence,
		})

		if confidence > bestConfidence {
			bestConfidence = confidence
			bestOutput = output
		}

		if confidence >= confidenceThreshold {
			return &ConfidenceResult{
				Output:       output,
				Confidence:   confidence,
				Attempts:     attempts,
				MetThreshold: true,
			}, nil
		}
	}

	return &ConfidenceResult{
		Output:       bestOutput,
		Confidence:   bestConfidence,
		Attempts:     attempts,
		MetThreshold: false,
	}, nil
}

func parseConfidenceResponse(text string) (string, float64) {
	confidence := 0.5

	confidenceRe := regexp.MustCompile(`(?i)CONFIDENCE:\s*([0-9.]+)`)
	if match := confidenceRe.FindStringSubmatch(text); len(match) > 1 {
		if conf, err := strconv.ParseFloat(match[1], 64); err == nil {
			if conf > 1.0 {
				conf = 1.0
			}
			if conf < 0.0 {
				conf = 0.0
			}
			confidence = conf
		}
	}

	output := confidenceRe.ReplaceAllString(text, "")
	output = strings.TrimSpace(output)

	return output, confidence
}

// ExampleWritingOptimization demonstrates the evaluator-optimizer pattern
func ExampleWritingOptimization() error {
	apiKey := getEnv("ANTHROPIC_API_KEY", "")
	if apiKey == "" {
		return fmt.Errorf("ANTHROPIC_API_KEY environment variable not set")
	}

	client := &AnthropicClient{
		APIKey:     apiKey,
		HTTPClient: nil, // Would use http.Client in production
	}

	optimizer := NewEvaluatorOptimizer(client, "claude-sonnet-4-20250514")

	// Add evaluation criteria
	optimizer.
		AddCriterion(EvaluationCriterion{
			Name:        "clarity",
			Description: "Writing is clear and easy to understand",
			Weight:      1.5,
		}).
		AddCriterion(EvaluationCriterion{
			Name:        "engagement",
			Description: "Content is engaging and holds attention",
			Weight:      1.2,
		}).
		AddCriterion(EvaluationCriterion{
			Name:        "accuracy",
			Description: "Information is accurate and well-researched",
			Weight:      1.5,
		})

	ctx := context.Background()
	result, err := optimizer.Optimize(
		ctx,
		"Write a blog post explaining how large language models work to a non-technical audience",
		3,
		0.85,
	)
	if err != nil {
		return err
	}

	fmt.Println("=== Optimization Results ===")
	fmt.Printf("Iterations: %d\n", result.Iterations)
	fmt.Printf("Final Score: %.1f%%\n", result.FinalScore*100)
	fmt.Printf("Met Threshold: %v\n", result.MetThreshold)

	fmt.Println("\n=== Score Progress ===")
	for _, iteration := range result.History {
		fmt.Printf("Iteration %d: %.1f%%\n", iteration.Iteration, iteration.Evaluation.OverallScore*100)
	}

	fmt.Printf("\n=== Final Output ===\n%s\n", result.FinalOutput)

	return nil
}

File: skills/agent-patterns/templates/rust/evaluator_optimizer.rs
/*!
 * Evaluator-Optimizer Pattern Implementation for Rust
 * Iterative refinement with generator and evaluator loop
 */

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

/// Anthropic API client
#[derive(Clone)]
pub struct AnthropicClient {
    api_key: String,
    http_client: reqwest::Client,
}

#[derive(Serialize)]
struct MessageRequest {
    model: String,
    max_tokens: u32,
    messages: Vec<MessageContent>,
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f64>,
}

#[derive(Serialize, Deserialize, Clone)]
struct MessageContent {
    role: String,
    content: String,
}

#[derive(Deserialize)]
struct MessageResponse {
    content: Vec<ContentBlock>,
}

#[derive(Deserialize)]
struct ContentBlock {
    #[serde(rename = "type")]
    content_type: String,
    text: Option<String>,
}

impl AnthropicClient {
    pub fn new(api_key: String) -> Self {
        Self {
            api_key,
            http_client: reqwest::Client::new(),
        }
    }

    pub async fn create_message(
        &self,
        prompt: &str,
        model: &str,
        max_tokens: u32,
        temperature: Option<f64>,
    ) -> Result<String> {
        let request = MessageRequest {
            model: model.to_string(),
            max_tokens,
            messages: vec![MessageContent {
                role: "user".to_string(),
                content: prompt.to_string(),
            }],
            temperature,
        };

        let response = self
            .http_client
            .post("https://api.anthropic.com/v1/messages")
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .header("content-type", "application/json")
            .json(&request)
            .send()
            .await?
            .json::<MessageResponse>()
            .await?;

        response
            .content
            .into_iter()
            .find(|c| c.content_type == "text")
            .and_then(|c| c.text)
            .context("No text content in response")
    }
}

/// Evaluation criteria with weight
#[derive(Debug, Clone)]
pub struct EvaluationCriterion {
    pub name: String,
    pub description: String,
    pub weight: f64,
}

/// Result of an evaluation
#[derive(Debug, Clone, Default)]
pub struct EvaluationResult {
    pub overall_score: f64,
    pub criteria_scores: std::collections::HashMap<String, f64>,
    pub feedback: String,
    pub suggestions: Vec<String>,
}

/// Iteration record for tracking progress
#[derive(Debug, Clone)]
pub struct IterationRecord {
    pub iteration: usize,
    pub output: String,
    pub evaluation: EvaluationResult,
}

/// Evaluator-Optimizer that iteratively refines output.
///
/// # Example
/// ```rust
/// let optimizer = EvaluatorOptimizer::new(client, "claude-sonnet-4-20250514".to_string());
/// optimizer.add_criterion(EvaluationCriterion {
///     name: "clarity".to_string(),
///     description: "Clear and understandable".to_string(),
///     weight: 1.5,
/// });
/// let result = optimizer.optimize("Write a blog post about AI", 3, 0.85).await?;
/// ```
pub struct EvaluatorOptimizer {
    client: Arc<AnthropicClient>,
    generator_model: String,
    evaluator_model: String,
    criteria: Vec<EvaluationCriterion>,
    history: Vec<IterationRecord>,
}

impl EvaluatorOptimizer {
    pub fn new(client: Arc<AnthropicClient>, model: String) -> Self {
        Self {
            client,
            generator_model: model.clone(),
            evaluator_model: model,
            criteria: Vec::new(),
            history: Vec::new(),
        }
    }

    pub fn with_evaluator_model(mut self, model: String) -> Self {
        self.evaluator_model = model;
        self
    }

    /// Add an evaluation criterion
    pub fn add_criterion(&mut self, criterion: EvaluationCriterion) -> &mut Self {
        self.criteria.push(criterion);
        self
    }

    /// Get iteration history
    pub fn history(&self) -> &[IterationRecord] {
        &self.history
    }

    /// Optimize output through iterative refinement
    pub async fn optimize(
        &mut self,
        task: &str,
        max_iterations: usize,
        score_threshold: f64,
    ) -> Result<OptimizationResult> {
        self.history.clear();
        let mut current_output = String::new();
        let mut last_evaluation: Option<EvaluationResult> = None;

        for i in 0..max_iterations {
            // Generate (or refine) output
            current_output = self
                .generate(task, &current_output, last_evaluation.as_ref())
                .await?;

            // Evaluate output
            let evaluation = self.evaluate(&current_output).await?;

            // Record iteration
            self.history.push(IterationRecord {
                iteration: i + 1,
                output: current_output.clone(),
                evaluation: evaluation.clone(),
            });

            // Check if we've met the threshold
            if evaluation.overall_score >= score_threshold {
                return Ok(OptimizationResult {
                    final_output: current_output,
                    final_score: evaluation.overall_score,
                    iterations: i + 1,
                    met_threshold: true,
                    history: self.history.clone(),
                });
            }

            last_evaluation = Some(evaluation);
        }

        // Return best result after max iterations
        let best = self
            .history
            .iter()
            .max_by(|a, b| {
                a.evaluation
                    .overall_score
                    .partial_cmp(&b.evaluation.overall_score)
                    .unwrap()
            })
            .unwrap();

        Ok(OptimizationResult {
            final_output: best.output.clone(),
            final_score: best.evaluation.overall_score,
            iterations: max_iterations,
            met_threshold: false,
            history: self.history.clone(),
        })
    }

    /// Generate or refine output
    async fn generate(
        &self,
        task: &str,
        previous_output: &str,
        previous_evaluation: Option<&EvaluationResult>,
    ) -> Result<String> {
        let prompt = if previous_output.is_empty() {
            format!(
                r#"Complete this task:

{}

Provide your best output:"#,
                task
            )
        } else {
            let feedback_text = if let Some(eval) = previous_evaluation {
                let suggestions: String = eval
                    .suggestions
                    .iter()
                    .map(|s| format!("- {}", s))
                    .collect::<Vec<_>>()
                    .join("\n");
                format!(
                    r#"Previous evaluation feedback:
{}

Specific suggestions:
{}"#,
                    eval.feedback, suggestions
                )
            } else {
                String::new()
            };

            format!(
                r#"Improve this output based on the feedback:

Original task: {}

Previous output:
{}

{}

Provide an improved version:"#,
                task, previous_output, feedback_text
            )
        };

        self.client
            .create_message(&prompt, &self.generator_model, 4096, None)
            .await
    }

    /// Evaluate output against criteria
    async fn evaluate(&self, output: &str) -> Result<EvaluationResult> {
        let criteria_list = if !self.criteria.is_empty() {
            self.criteria
                .iter()
                .map(|c| format!("- {} (weight: {}): {}", c.name, c.weight, c.description))
                .collect::<Vec<_>>()
                .join("\n")
        } else {
            "- quality: Overall quality and correctness\n\
             - clarity: Clear and understandable\n\
             - completeness: Addresses all aspects"
                .to_string()
        };

        let prompt = format!(
            r#"Evaluate this output against the following criteria:

{}

Output to evaluate:
{}

Respond with JSON in this exact format:
{{
    "overall_score": 0.0-1.0,
    "criteria_scores": {{
        "criterion_name": 0.0-1.0
    }},
    "feedback": "Overall assessment",
    "suggestions": ["specific improvement 1", "specific improvement 2"]
}}"#,
            criteria_list, output
        );

        let response = self
            .client
            .create_message(&prompt, &self.evaluator_model, 1024, None)
            .await?;

        parse_evaluation_json(&response)
    }
}

fn parse_evaluation_json(json: &str) -> Result<EvaluationResult> {
    let mut result = EvaluationResult::default();

    // Extract overall score
    if let Some(score) = extract_json_number(json, "overall_score") {
        result.overall_score = score;
    }

    // Extract feedback
    if let Some(feedback) = extract_json_string(json, "feedback") {
        result.feedback = feedback;
    }

    // Extract suggestions
    if let Some(suggestions_match) = regex::Regex::new(r#""suggestions"\s*:\s*\[(.*?)\]"#)
        .ok()
        .and_then(|re| re.captures(json))
    {
        let suggestions_str = &suggestions_match[1];
        result.suggestions = regex::Regex::new(r#""([^"]+)""#)
            .ok()
            .map(|re| {
                re.captures_iter(suggestions_str)
                    .map(|c| c[1].to_string())
                    .collect()
            })
            .unwrap_or_default();
    }

    Ok(result)
}

fn extract_json_string(json: &str, key: &str) -> Option<String> {
    let pattern = format!(r#""{}":\s*"([^"]*)""#, key);
    regex::Regex::new(&pattern)
        .ok()?
        .captures(json)
        .map(|c| c[1].to_string())
}

fn extract_json_number(json: &str, key: &str) -> Option<f64> {
    let pattern = format!(r#""{}":\s*([0-9.]+)"#, key);
    regex::Regex::new(&pattern)
        .ok()?
        .captures(json)
        .and_then(|c| c[1].parse().ok())
}

#[derive(Debug)]
pub struct OptimizationResult {
    pub final_output: String,
    pub final_score: f64,
    pub iterations: usize,
    pub met_threshold: bool,
    pub history: Vec<IterationRecord>,
}

/// Confidence-based optimizer
pub struct ConfidenceBasedOptimizer {
    client: Arc<AnthropicClient>,
    model: String,
}

impl ConfidenceBasedOptimizer {
    pub fn new(client: Arc<AnthropicClient>, model: String) -> Self {
        Self { client, model }
    }

    /// Generate with confidence self-assessment
    pub async fn generate_with_confidence(
        &self,
        task: &str,
        confidence_threshold: f64,
        max_attempts: usize,
    ) -> Result<ConfidenceResult> {
        let mut attempts = Vec::new();
        let mut best_output = String::new();
        let mut best_confidence = 0.0;

        for i in 0..max_attempts {
            let prompt = format!(
                r#"Complete this task and assess your confidence:

{}

After your response, on a new line, provide your confidence level (0.0-1.0) that your answer is correct and complete.

Format:
[Your response here]

CONFIDENCE: [0.0-1.0]"#,
                task
            );

            let temperature = if i == 0 { None } else { Some(0.3) };
            let response = self
                .client
                .create_message(&prompt, &self.model, 4096, temperature)
                .await?;

            let (output, confidence) = parse_confidence_response(&response);

            attempts.push(AttemptRecord {
                attempt: i + 1,
                output: output.clone(),
                confidence,
            });

            if confidence > best_confidence {
                best_confidence = confidence;
                best_output = output.clone();
            }

            if confidence >= confidence_threshold {
                return Ok(ConfidenceResult {
                    output,
                    confidence,
                    attempts,
                    met_threshold: true,
                });
            }
        }

        Ok(ConfidenceResult {
            output: best_output,
            confidence: best_confidence,
            attempts,
            met_threshold: false,
        })
    }
}

fn parse_confidence_response(text: &str) -> (String, f64) {
    let confidence = regex::Regex::new(r"(?i)CONFIDENCE:\s*([0-9.]+)")
        .ok()
        .and_then(|re| re.captures(text))
        .and_then(|c| c[1].parse::<f64>().ok())
        .map(|v| v.clamp(0.0, 1.0))
        .unwrap_or(0.5);

    let output = regex::Regex::new(r"(?i)\nCONFIDENCE:\s*[0-9.]+")
        .ok()
        .map(|re| re.replace_all(text, "").trim().to_string())
        .unwrap_or_else(|| text.to_string());

    (output, confidence)
}

#[derive(Debug)]
pub struct ConfidenceResult {
    pub output: String,
    pub confidence: f64,
    pub attempts: Vec<AttemptRecord>,
    pub met_threshold: bool,
}

#[derive(Debug)]
pub struct AttemptRecord {
    pub attempt: usize,
    pub output: String,
    pub confidence: f64,
}

// Example usage
pub async fn example_writing_optimization() -> Result<()> {
    let api_key = std::env::var("ANTHROPIC_API_KEY")
        .context("ANTHROPIC_API_KEY environment variable not set")?;

    let client = Arc::new(AnthropicClient::new(api_key));
    let mut optimizer = EvaluatorOptimizer::new(client, "claude-sonnet-4-20250514".to_string());

    // Add evaluation criteria
    optimizer
        .add_criterion(EvaluationCriterion {
            name: "clarity".to_string(),
            description: "Writing is clear and easy to understand".to_string(),
            weight: 1.5,
        })
        .add_criterion(EvaluationCriterion {
            name: "engagement".to_string(),
            description: "Content is engaging and holds attention".to_string(),
            weight: 1.2,
        })
        .add_criterion(EvaluationCriterion {
            name: "accuracy".to_string(),
            description: "Information is accurate and well-researched".to_string(),
            weight: 1.5,
        });

    let result = optimizer
        .optimize(
            "Write a blog post explaining how large language models work to a non-technical audience",
            3,
            0.85,
        )
        .await?;

    println!("=== Optimization Results ===");
    println!("Iterations: {}", result.iterations);
    println!("Final Score: {:.1}%", result.final_score * 100.0);
    println!("Met Threshold: {}", result.met_threshold);

    println!("\n=== Score Progress ===");
    for iteration in &result.history {
        println!(
            "Iteration {}: {:.1}%",
            iteration.iteration,
            iteration.evaluation.overall_score * 100.0
        );
    }

    println!("\n=== Final Output ===\n{}", result.final_output);

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_confidence() {
        let text = "Here is my answer.\n\nCONFIDENCE: 0.85";
        let (output, confidence) = parse_confidence_response(text);
        assert!(!output.contains("CONFIDENCE"));
        assert!((confidence - 0.85).abs() < 0.01);
    }

    #[test]
    fn test_evaluation_criterion() {
        let criterion = EvaluationCriterion {
            name: "test".to_string(),
            description: "Test criterion".to_string(),
            weight: 1.0,
        };
        assert_eq!(criterion.name, "test");
    }
}
